RUBY 2704 wex charging bo 3 1 3 refund a payment page

jjromeo opened the pull request:
- [RUBY-3414] Add payment details page and link to registrations
- [RUBY-3414] Add conditional display for balance in resource details view
- [RUBY-3414] Add safe navigation operator to account balance in resource details view
- [RUBY-3414] Add request specs for payment details view
- [RUBY-3414] Remove `companies` table from schema.rb
- [RUBY-3414] Remove unused `view_certificate_token` and `view_certificate_token_created_at` columns from schema.
- [RUBY-2704] Add refund payment functionality
- [RUBY-2704] Add 'Refund a Payment' feature to the Waste Exemptions back-office
- [RUBY-2704] Rubocop fixes


[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-3414]: https://eaflood.atlassian.net/browse/RUBY-3414?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-2704]: https://eaflood.atlassian.net/browse/RUBY-2704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-2704]: https://eaflood.atlassian.net/browse/RUBY-2704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RUBY-2704]: https://eaflood.atlassian.net/browse/RUBY-2704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jjromeo committed:
[RUBY-3414] Add payment details page and link to registrations

- Created `PaymentDetailsController` with an `index` action to display payment details.
- Added a new view `index.html.erb` for the payment details page.
- Updated `RegistrationsController` to remove the `authorize` method.
- Enhanced `ActionLinksHelper` with `display_payment_details_link_for?` method.
- Modified `_resource_details.html.erb` to include current balance and link to payment details.
- Updated locale files

jjromeo committed:
[RUBY-3414] Add conditional display for balance in resource details view

Wrapped the balance display section in a conditional to only show if the `resource` is a `WasteExemptionsEngine::Registration`.

jjromeo committed:
[RUBY-3414] Add safe navigation operator to account balance in resource details view

This change ensures that the `balance` is accessed safely using the safe navigation operator to prevent potential nil errors when `account` is nil.

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Add request specs for payment details view

- Created `payment_details_spec.rb` to test the payment details page.
- Includes tests for rendering the show template with correct registration reference when a user is signed in.
- Includes tests for redirecting to the sign-in page when no valid user is signed in.

jjromeo committed:
Merge branch 'RUBY-3414-wex-charging-bo-1-2-view-details-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Remove `companies` table from schema.rb

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
Merge branch 'main' into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-3414] Remove unused `view_certificate_token` and `view_certificate_token_created_at` columns from schema.

jjromeo committed:
Merge branch 'RUBY-3414-wex-charging-bo-1-2-view-details-page' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-3414-wex-charging-bo-1-2-view-details-page

jjromeo committed:
[RUBY-2704] Add refund payment functionality

- Implement `RecordRefundFormsController` to handle refund operations, including `index`, `new`, and `create` actions.
- Create `RecordRefundForm` model with validation and submission logic for recording refunds.
- Introduce `RecordRefundService` to process refunds and update account balances.
- Add views for listing payments eligible for refunds and for creating a refund.
- Update `FinanceDetailsHelper` to adjust currency display logic.
- Modify routes to include refund paths under

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Add 'Refund a Payment' feature to the Waste Exemptions back-office

- Updated `Gemfile` and `Gemfile.lock` to point to the `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` branch of the `waste_exemptions_engine` for the new refund feature.
- Modified `payment_details_controller.rb` to use a new `PaymentDetailsPresenter` for displaying payment details.
- Enhanced `record_refund_form.rb` to include validation for refund amounts, ensuring they do not exceed the payment amount or account balance.
- Created `PaymentDetailsPresenter` to format and present payment, order, and refund details in the views.
- Refactored `payment_details/index.html.erb` to use the presenter for displaying payment details, including orders, payments, refunds, and balance.
- Updated `record_refund_forms/new.html.erb` to use the GOV.UK Design System FormBuilder and display error summaries.
- Added new locale entries in `record_refund_forms.en.yml` and `payment_details.en.yml` for improved internationalization support, including error messages for refund validation.
- Adjusted the database schema to include a `comments` field in the `payments` table and re-ordered some indexes for efficiency.
- Updated `record_refund_forms_spec.rb` to reflect changes in the refund form logic and remove outdated comments.
- This commit introduces a comprehensive refund processing feature, enhancing the back-office capabilities for managing payments and refunds efficiently.

jjromeo committed:
[RUBY-2704] Rubocop fixes

- Refactor `RecordRefundFormsController` to improve code readability and ensure correct method chaining.
- Update `RecordRefundForm` to include validations and refactor methods for better clarity.
- Add localization for success and error messages in `record_refund_forms.en.yml`.
- Use `delegate` in `PaymentDetailsPresenter` for cleaner code and update method chaining.
- Adjust formatting in `FinanceDetailsHelper`.
- Modify request specs in `record_refund_forms_spec

jjromeo committed:
Add `i18n_page` variable to payment details spec

Added the `i18n_page` variable to the `GET /registrations/:reference/payment_details` test context to improve clarity and maintainability of internationalization references.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForms` to `RecordRefunds` for consistency

Renamed the `RecordRefundFormsController` to `RecordRefundsController` and updated all related references throughout the application. This includes changes in the controller, views, routes, localization files, and request specs. Adjusted the I18n keys and paths to reflect the new naming convention, ensuring consistency and clarity across the codebase.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Refactor `reason` to `comments` in refund functionality

Changed the attribute `reason` to `comments` across various files for consistency in refund processing. Updated models, controllers, services, views, and locale files to reflect this change. Adjusted validation and logging accordingly.

jjromeo committed:
[RUBY-2704] Rename `record_refund_forms_controller.rb` to `record_refunds_controller.rb` for consistency

jjromeo committed:
[RUBY-2704] Refactor refund form and update localization keys

- Updated error message keys in `RecordRefundForm` to use a more consistent path.
- Changed the path for new refund forms in views and specs from `new_registration_record_refund_form_path` to `new_registration_record_refund_path`.
- Updated localization files to rename `reason` to `comments` and adjusted error message paths.
- Removed unused `account_id` attribute from the payment factory.
- Updated specs to reflect changes in path and parameter

jjromeo committed:
[RUBY-2704] Update 'comments' to 'reason' in payment details localization and specs

- Changed the label 'comments' to 'reason' in `payment_details.en.yml`.
- Updated corresponding test expectations in `payment_details_spec.rb` to reflect this change.

jjromeo committed:
[RUBY-2704] Remove bang method from `update_balance` call in `RecordRefundService` to prevent raising exceptions.

jjromeo committed:
[RUBY-2704] Remove schema changes as they have been handled already

jjromeo committed:
[RUBY-2704] Update `Gemfile.lock` to latest `waste-exemptions-engine` revision

Updated the `waste-exemptions-engine` gem to revision `aa541fd649a174b96ebfc1466ed7d710fc44e9f0` in the `Gemfile.lock` to incorporate the latest changes on the `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` branch.

jjromeo committed:
[RUBY-2704] Add refund functionality enhancements

- Updated `Gemfile.lock` to reflect the latest revision of the waste-exemptions-engine.
- Modified `RecordRefundForm` to include validation for comments presence, ensuring a reason is provided for refunds.
- Added `can_display_refund_link?` method to `PaymentDetailsPresenter` to conditionally display refund links based on account status.
- Enhanced `RecordRefundService` to append "/REFUND" to the payment reference for clarity.
- Updated view logic in

jjromeo committed:
[RUBY-2704] Enhance refund feature and update payment display

- Added `reason_present_in_comments` validation to `RecordRefundForm`.
- Changed date formatting in `PaymentDetailsPresenter` to use `:day_month_year_slashes`.
- Updated payment and refund tables in views to use `created_at` instead of `date_time`.
- Replaced `order_id` with `reference` in the refund index view.
- Displayed payment comments instead of payment type in the refund table.

jjromeo committed:
[RUBY-2704] Add tests for `RecordRefundService` to handle refund processing

This commit introduces a new test suite for the `RecordRefundService` to verify its functionality when processing refunds. It includes tests for successful refund creation, correct attribute assignment, account balance updates, and error handling with logging and Airbrake notification. The tests ensure that the service behaves as expected in both successful and error scenarios.

jjromeo committed:
[RUBY-2704] Update `@payments` to use `refundable` scope in `RecordRefundsController#index` method.

jjromeo committed:
[RUBY-2704] Refactor `validate_amount` logic in `RecordRefundForm`

Simplified the condition for adding an error when the refund amount exceeds the balance by removing unnecessary checks and directly comparing the amount to the balance.

jjromeo committed:
[RUBY-2704] Enhance refund form validation and add tests

- Added `payment_exists` validation to ensure the payment ID is valid and exists.
- Improved the `submit` method to handle cases where the payment is not found.
- Updated locale file to include error message for missing payment.
- Added RSpec tests for `RecordRefundForm` to validate various scenarios, including invalid amounts, nonexistent payments, and missing comments.

jjromeo committed:
Merge branch 'main' of https://github.com/DEFRA/waste-exemptions-back-office into RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page

jjromeo committed:
[RUBY-2704] Refactor refund process and update payment details view

- Updated `Gemfile.lock` with the latest `waste_exemptions_engine` revision.
- Removed account balance update logic from `RecordRefundService` and corresponding test case.
- Refactored `payment_details/index.html.erb` to reorganize sections for charges, payments, and balances.
- Improved table structure and translation keys to enhance clarity and maintainability.

jjromeo committed:
[RUBY-2704] Add conditional display for refund link in payment details view

Wrapped the refund link in a conditional check using `@presenter.can_display_refund_link?` to control its visibility based on specific conditions.

brujeo commented on app/controllers/record_refunds_controller.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def authorize
+    authorize! :read, @resource
+  end
```
I believe it was agreed that these payment pages would not have any permission checks initially. Permission checks will be added later once appropriate user roles are defined.

brujeo commented on app/forms/record_refund_form.rb:
```diff
@@ -0,0 +1,60 @@
+
+  attr_accessor :payment_id, :amount, :comments
+
+  validates :amount, presence: true, numericality: { greater_than: 0 }
```
I think Tim won't pass it through as this allows any numbers to be added (e.g. 234.46234523425). But it has to check for max 2 decimal points. For payment amount I had to create a dedicated validator in defra_ruby_validators gem

 validates :payment_amount, "defra_ruby/validators/price": true

brujeo replied:
Another issue with accepting amounts with more than 2 decimal point, it will result in incorrect (rounded) value being  stored during conversion from pounds to pence, e.g. amount of 23.4678 will result in 2346.78 pence but we store only integers in db

brujeo commented on app/views/record_refunds/new.html.erb:
```diff
@@ -0,0 +1,52 @@
+      <div class="govuk-form-group">
+        <h2 class="govuk-heading-m"><%= t(".maximum_refund_amount") %></h2>
+        <p class="govuk-body">
+          £<%= display_pence_as_pounds_and_cents(@payment.payment_amount, hide_pence_if_zero: false)%>
+
```
Why not to use PaymentDetailsPresenter here? Actually, it would be good to change other payment-related pages as well and make use of this presenter

jjromeo replied:
Generally I think this is a good idea, I am just wondering though in this case what the advantage would be? to reduce reliance on the helpers?

brujeo replied:
the obvious benefit would be that we have the same display rules for the value on all pages. Currently we have you using this format
`£<%= display_pence_as_pounds_and_cents(@payment.payment, hide_pence_if_zero: false)%>`
me using 
`number_to_currency(display_pence_as_pounds_and_cents(payment.payment), unit: "£")`
and possibly Paul using something different

jjromeo committed:
[RUBY-2704] Update `RecordRefundForm` validation for `amount`

Replaced the `amount` validation with a custom `defra_ruby/validators/price` validator to ensure more accurate validation logic.

jjromeo committed:
[RUBY-2704] Add validation for refund amount in `RecordRefundForm`

- Ensure refund `amount` is present and greater than 0.
- Update tests to check for negative amounts and high precision amounts, ensuring they return appropriate error messages.

jjromeo committed:
[RUBY-2704] Update `waste_exemptions_engine` gem to use the `main` branch

Switched the branch for the `waste_exemptions_engine` gem in the `Gemfile` and `Gemfile.lock` from `RUBY-2704-wex-charging-bo-3-1-3-refund-a-payment-page` to `main`. Updated the revision to reflect the latest commit on the `main` branch.

brujeo commented on app/forms/record_refund_form.rb:
```diff
@@ -6,6 +6,7 @@ class RecordRefundForm
   attr_accessor :payment_id, :amount, :comments
 
   validates :amount, presence: true, numericality: { greater_than: 0 }
+  validates :amount, "defra_ruby/validators/price": true
```
You may need to specify your custom error messages for price validator, e.g. 
`validates :payment_amount,
            "defra_ruby/validators/price": { messages: {
              blank: I18n.t(".payments.errors.payment_amount_blank"),
              invalid_format: I18n.t(".payments.errors.payment_amount_invalid")
            } }`

jjromeo committed:
[RUBY-2704] Remove unused `authorize` method from `RecordRefundsController`

The `authorize` method was removed as it was not being called in the controller. Additionally, a commented-out `authorize` call in the `index` method was also deleted to clean up the code.

jjromeo committed:
[RUBY-2704] Improve validation message for refund amount

Updated `RecordRefundForm` to provide a clearer error message when the refund amount is zero, ensuring users receive a more descriptive prompt. Adjusted corresponding spec to match the updated validation message.

jjromeo committed:
[RUBY-2704] Add localized error messages for refund amount validation

- Updated `RecordRefundForm` to use I18n for error messages when the refund amount is blank or zero.
- Added new error messages to `record_refunds.en.yml` for better localization support.
- Adjusted specs in `record_refund_form_spec.rb` to reflect the updated error messages and added a test case for a blank amount.

jjromeo committed:
[RUBY-2704] Refactor `RecordRefundForm` validation formatting

Adjusted the indentation for the `amount` validation rules in `RecordRefundForm` for better readability and consistency.

jjromeo committed:
[RUBY-2704] Add price validation to `amount` in `RecordRefundForm`

Moved the price validation for `amount` to be grouped with other validations. This ensures the `amount` field is validated for presence and numericality, and adheres to the price format.
